src/kafka-handler.ts

```diff
@@ -68,29 +58,75 @@ class KafkaHandler implements DataBroker {
-      client = new kafka.Client(config.broker.host, "iotagent-json-" +  Math.floor(Math.random() * 10000));
-      console.log("... Kafka client was created.");
+  private initKafkaConfiguration(config: config.ConfigOptions, callback: (data: DeviceManagerEvent) => void, type: string) {
+    switch (type) {
+      case "producer":
+        if (config.broker.type === "kafka") {
+          console.log("Creating Kafka producer...");
+          let client = new kafka.Client(config.broker.host, "iotagent-json-producer-" +  Math.floor(Math.random() * 10000));
+          console.log("... Kafka client for producer is ready.");
+          console.log("Creating Kafka producer...");
+          this.producer = new kafka.HighLevelProducer(client, { requireAcks: 1 });
+      
+          // Kafka producer events registration
+          this.producer.on("ready", () => {
+            console.log("... Kafka producer creation finished.");
+            this.isProducerReady = true;
+            this.initKafkaConfiguration(config, callback, "consumer");
+          });
+          this.producer.on("error", (e) => { 
+            console.error("Error: ", e); 
+            console.log("Will try again in a few seconds.");
+            setTimeout(() => {
+              console.log("Trying again.");
+              this.initKafkaConfiguration(config, callback, "producer");
+            }, 1000)
+          });
+          console.log("... Kafka producer created and callbacks registered.");
+        } else {
+          console.log("Data broker is not Kafka. Skipping producer creation.");
+          this.isProducerReady = true;
+          this.initKafkaConfiguration(config, callback, "consumer");
+        }
+        break;
+      case "consumer":
+        console.log("Creating new Kafka client...");
+        // First try
+        // Creating consumer client - from device manager to iotagent.
+        // This is always used.
+        let client = new kafka.Client(config.device_manager.kafkaHost, "iotagent-json-consumer-" +  Math.floor(Math.random() * 10000));
+        console.log("... Kafka client for consumer is ready.");
+        console.log("Creating Kafka consumer...");
+        this.consumer = new kafka.HighLevelConsumer(client, config.device_manager.kafkaTopics, config.device_manager.kafkaOptions);
+        // Kafka consumer events registration
+        this.consumer.on("message", (data) => {
+          let parsedData = JSON.parse(data.value.toString());
+          callback(parsedData);
+        });
+        this.consumer.on("error", (err) => {
+            console.log("Error: ", err);
+            console.log("Will try again in a few seconds.");
+            setTimeout(() => {
+              console.log("Trying again.");
+              this.initKafkaConfiguration(config, callback, "consumer");
+            }, 1000)
+        });
+
+        console.log("... Kafka consumer created and callbacks registered.");
+        this.initKafkaConfiguration(config, callback, "finish");
```

Comment:
Why isn't this part a continuous sequence of code? Why the case 'finish': and recalling the function to reach that part of the code?

Reply:
No need for the 'finish' case.

Reply:
There is no need for *a* finish case. No need to create topics.